File: 3-longestSubstring/Cpp/longestSubstring.cpp
#include <map>
#include <string>
#include <iostream>
#include <minmax.h>

class Solution
{
public:
  int lengthOfLongestSubstring(std::string s)
  {
    int longest = 0;
    auto sz = s.size();
    auto seen_map = std::map<char, int>{};
    
    for (auto i = 0; i < sz; i++)
    {
      auto ch = s[i];
      if(seen_map.contains(ch)){
        if (auto sms = seen_map.size(); longest < sms)
        {
          longest = sms;
        }
        auto idx = seen_map[ch];

        for (auto it = seen_map.begin(); it != seen_map.end();)
        {
          if (it->second <= idx)
            it = seen_map.erase(it);
          else
            ++it;
        }
      }
      seen_map[ch] = i;
    }
    if (auto sms = seen_map.size(); longest < sms)
    {
      longest = sms;
    }

    return longest;
  }
  int lolss10xfaster(std::string s)
  {
    int len = s.size();
    int char_map[127] = { 0 };
    int longest = 0;
    for (int left = 0, right = 0; right < len; right++)
    {
      left = max(char_map[s[right]], left);
      longest = max(longest, right - left + 1);
      char_map[s[right]] = right + 1;
    }
    return longest;
  }
};

int main()
{
  std::cout << Solution().lolss10xfaster("dvdf");
}

File: 3-longestSubstring/JS/longestSubstring.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function (s) {
  //   if (s === "") return 0;
  //   return 1;

  let left = 0;
  let mx = 0;
  const char_map = new Map();

  for (let right = 0; right < s.length; right++) {
    ch = s.at(right);
    lch = char_map.get(ch) ? char_map.get(ch) : 0;
    left = Math.max(lch, left);
    mx = Math.max(mx, right - left + 1);
    char_map.set(ch, right + 1);
  }

  return mx;
};

module.exports = lengthOfLongestSubstring;
